### ISSUE TITLE
Placeholder outside of CMS Permissions issue

### ISSUE THREAD

**Issue by jtiki** [divio/django-cms#5178]
I am building an application that uses the PlaceholderField in a model, but cannot seem to get the permissions working correctly. According to the documentation: 

> Permissions
> To be able to edit a placeholder user must be a staff member and needs either edit permissions on the model that contains the PlaceholderField, or permissions for that specific instance of that model. Required permissions for edit actions are:
> 
> * to add: require add or change permission on related Model or instance.
> * to change: require add or change permission on related Model or instance.
> * to delete: require add or change or delete permission on related Model or instance.

So I built my model: 
```
class UserProfile(models.Model):
    user = models.OneToOneField(User, related_name='user_profile')
    address = models.CharField(max_length=55, blank=True, null=True,
                                                   help_text="Home Address, ex: 111 Street Addr")
     address2 = models.CharField(max_length=55, blank=True, null=True,
                                                      help_text="City, State Zip Code")
    phone = models.CharField(max_length=12, blank=True, null=True)
    facebook = models.URLField(blank=True, null=True)
    twitter = models.URLField(blank=True, null=True)
    google_plus = models.URLField(blank=True, null=True)
    user_info_placeholder = PlaceholderField('UserInfoPlaceholder',
                                                                         related_name='userinfoplaceholder')
    timeline = PlaceholderField('TimelinePlaceholder', related_name='timeline')
```
Added a view:
```
class UserProfileDetailView(DetailView):
    context_object_name = 'user_information'
    model = User
    slug_field = 'username'
    template_name = 'home_page.html'
```
And logged in as the user, however no 'Structure/Content' button appeared on the toolbar. To check I logged in as a superuser and the button showed. So I changed the users "User Permissions" and added everything just to be sure, but still not 'Structure/Content' button.

Any suggestions?

**Comment by jtiki** [comment]
I should also note the user is 'active' and has 'staff' status.

**Comment by yakky** [comment]
@jtiki you have to add the "Can use structure mode" permission to the user

**Comment by jtiki** [comment]
@yakky that would be in the M2M field when editing a user right? I added all permissions (using the "Choose all" link) and it didn't seem to have any affect.

**Comment by jtiki** [comment]
I went back to my install and checked it appears I do have "Can use structure mode" under the users permission:

<img width="1042" alt="screen shot 2016-04-19 at 6 46 04 pm" src="https://cloud.githubusercontent.com/assets/1173792/14658028/15d5efe6-065f-11e6-8bf7-c9d698d49486.png">

**Comment by czpython** [comment]
I believe this is both a validation and documentation problem.
I can reproduce the issue only if the admin class for the object that has the `PlaceholderField` does not subclass from `PlaceholderAdminMixin`.

Will open a new ticket for the documentation issue and close this one for https://github.com/divio/django-cms/issues/3734

**Comment by jtiki** [comment]
@czpython I just checked this solution in my code and that was the problem. Once I added PlaceholderAdminMixin everything started working! Thanks for the fix!

**Comment by openHBP** [comment]
I see this issue is closed and is the same as #6153 and #5410 but I miss something...

In my `msg.models.py`, I subclass the PlaceholderAdminMixin like this:

```python
from django.db import models
from cms.models.fields import PlaceholderField
from cms.admin.placeholderadmin import PlaceholderAdminMixin

class MsgDet(PlaceholderAdminMixin, models.Model):
    title = models.CharField(max_length=255)
    short_text = PlaceholderField('msgdet_slot')
```
I've also tried:
```python
from django.db import models
from cms.models.fields import PlaceholderField
from cms.admin.placeholderadmin import PlaceholderAdminMixin

class MyPlaceholderField(PlaceholderAdminMixin, PlaceholderField):
    pass

class MsgDet(models.Model):
    title = models.CharField(max_length=255)
    short_text = MyPlaceholderField('msgdet_slot')
```

None of them are permitting access to my user on the placeholder slot defined as PlaceholderField.

- I've done `makemigration` and `migrate` on my model.
- I've granted `all permissions` to my user.

I always get the same message after double clicking on the `rendered PlaceholderField` on my html page: _"You don't have permission to modify this plugin"_.
What I'm missing here???

### Environment
Python==3.6.9
Django==3.0
django-cms==3.7.4
django-polymorphic==2.1.2

**Comment by openHBP** [comment]
to solve this issue, I've removed the fact that there must be an attached model in `cms.models.placeholdermodel.Placeholder`

```py
    def has_change_permission(self, user):
        """
        Returns True if user has permission
        to change all models attached to this placeholder.
        """
        from cms.utils.permissions import get_model_permission_codename

        attached_models = self._get_attached_models()
        
        # Comment this part!
        # if not attached_models:
        #     # technically if placeholder is not attached to anything,
        #     # user should not be able to change it but if is superuser
        #     # then we "should" allow it.
        #     return user.is_superuser

        attached_objects = self._get_attached_objects()

        for obj in attached_objects:
            try:
                perm = obj.has_placeholder_change_permission(user)
            except AttributeError:
                model = type(obj)
                change_perm = get_model_permission_codename(model, 'change')
                perm = user.has_perm(change_perm)

            if not perm:
                return False
        return True
```
